src/test/java/com/teragrep/rlp_03/client/ClientTest.java

```diff
@@ -0,0 +1,222 @@
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class ClientTest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClientTest.class);
+
+    private final String hostname = "localhost";
+    private Server server;
+    private final int port = 23601;
+
+    @BeforeAll
+    public void init() {
+        Config config = new Config(port, 1);
+        ServerFactory serverFactory = new ServerFactory(
+                config,
+                () -> new DefaultFrameDelegate((frame) -> LOGGER.debug("server got <[{}]>", frame.relpFrame()))
+        );
+        Assertions.assertAll(() -> {
+            server = serverFactory.create();
+
+            Thread serverThread = new Thread(server);
+            serverThread.start();
+
+            server.startup.waitForCompletion();
+        });
+    }
+
+    @Test
+    public void t() throws IOException, InterruptedException {
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        SocketFactory socketFactory = new PlainFactory();
+
+        ConnectContextFactory connectContextFactory = new ConnectContextFactory(executorService, socketFactory);
+
+        // this is for returning ready connection
+        CompletableFuture<ConnectionContext> readyContextFuture = new CompletableFuture<>();
+
+        Consumer<ConnectionContext> connectionContextConsumer = connectionContext -> {
+            LOGGER.debug("connectionContext ready");
+            readyContextFuture.complete(connectionContext);
+        };
+
+        // TODO perhaps <Integer, Future<Something>> ?
+        // FIXME what should be Something, RelpFrame is immediately deallocated after FrameDelegate
+        // TODO design better: Futures are not optimal for multi-complete/disruptor pattern
+        HashMap<Integer, CompletableFuture<String>> pendingTransactions = new HashMap<>();
+
+        FrameDelegate essentialClientDelegate = new FrameDelegate() {
+
+            @Override
+            public boolean accept(FrameContext frameContext) {
+                LOGGER.debug("client got <[{}]>", frameContext.relpFrame());
+
+                int txn = frameContext.relpFrame().txn().toInt();
+
+                // TODO implement better handling for hint frames
+                if (txn == 0) {
+                    return true;
+                }
+
+                CompletableFuture<String> future = pendingTransactions.remove(txn);
+
+                if (future == null) {
+                    throw new IllegalStateException("txn not pending <[" + txn + "]>");
+                }
```

Comment:
Assert instead

Reply:
exception is part of the logic required, meaning it is not for testing. futures are tested elsewhere.